### PR TITLE
executor, store: Plumb the query max execution time to timebox the PD GetRegion grpc calls

### PR DESCRIPTION
**Pull request by HaoW30**
### What problem does this PR solve?
This change plumbs the TiDB query max execution time to the executor point_get(bath_point_get) and coprocessor(batch_coprocessor) tasks to timebox the GetRegion PD grpc calls (and other task operations). 

Issue Number: ref #56753

Problem Summary:

### What changed and how does it work?
1. Plumbed query execution time to kv request and distsql context
2. Set context deadline based on query max execution time for point get snapshot read (point_get and batch_point_get)
3. Set context deadline based on query max execution time for coprocessor tasks (coprocessor and batch_coprocessor)

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by lhy1024 — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
lgtm

**Review by bb7133 — APPROVED** [review]
LGTM

File: pkg/executor/batch_point_get.go [hunk 1]
```diff
@@ -235,6 +236,10 @@ func (e *BatchPointGetExec) initialize(ctx context.Context) error {
 	var indexKeys []kv.Key
 	var err error
 	batchGetter := e.batchGetter
+	if e.Ctx().GetSessionVars().MaxExecutionTime > 0 {
+		// If MaxExecutionTime is set, we need to set the context deadline for the batch get.
+		ctx, _ = context.WithTimeout(ctx, time.Duration(e.Ctx().GetSessionVars().MaxExecutionTime)*time.Millisecond)
```
Comment (cfzjywxk) [inline pkg/executor/batch_point_get.go:241]: It is safer to to use
```
ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
defer cancel()
```
to follow the best practise and avoid unexpected resource usage risks.

File: pkg/distsql/request_builder.go [hunk 2]
```diff
@@ -341,6 +341,7 @@ func (builder *RequestBuilder) SetFromSessionVars(dctx *distsqlctx.DistSQLContex
 	builder.Request.StoreBusyThreshold = dctx.LoadBasedReplicaReadThreshold
 	builder.Request.RunawayChecker = dctx.RunawayChecker
 	builder.Request.TiKVClientReadTimeout = dctx.TiKVClientReadTimeout
+	builder.Request.MaxExecutionTime = dctx.MaxExecutionTime
```
Comment (cfzjywxk) [inline pkg/distsql/request_builder.go:344]: IMO adding a timeout to the context should be the final step. If we design with the expectation that the entire execution path checks for `ctx.Done` messages, perhaps we should review and document in the gitHub issue which tidb paths lack `ctx.Done` checks and list related TODO sub-tasks.

/cc @bb7133

File: pkg/executor/point_get.go [hunk 3]
```diff
@@ -672,7 +673,13 @@ func (e *PointGetExecutor) get(ctx context.Context, key kv.Key) ([]byte, error)
 		}
 	}
 	// if not read lock or table was unlock then snapshot get
-	return e.snapshot.Get(ctx, key)
+	if e.Ctx().GetSessionVars().MaxExecutionTime > 0 {
+		// if the query has max execution time set, we need to set the context deadline for the get request
+		ctxWithTimeout, _ := context.WithTimeout(ctx, time.Duration(e.Ctx().GetSessionVars().MaxExecutionTime)*time.Millisecond)
```
Comment (cfzjywxk) [inline pkg/executor/point_get.go:678]: Ditto.

File: pkg/executor/batch_point_get.go (hunk 1 above)
Comment (HaoW30) [reply pkg/executor/batch_point_get.go:241]: yes, will do to avoid context leakage

**Pull request merged**